tft_bot/config.py

```diff
@@ -0,0 +1,112 @@
+"""
+Module to handle the configuration for the bot.
+"""
+import argparse
+import os.path
+import shutil
+from typing import Any
+
+from loguru import logger
+from ruamel.yaml import YAML
+
+from .helpers import system_helpers
+
+_SELF: dict[str, Any] = {}
+_CLI_FLAGS = {}
+
+
+def load_config(storage_path: str) -> None:
+    """
+    Writes the configuration from resource (provided in repository) to storage path, loads the configuration from
+    storage path to memory and updates the configuration if necessary.
+
+    Args:
+        storage_path: The base storage path where all of our files should go.
+
+    """
+    yaml = YAML()
+
+    config_resource_path = system_helpers.resource_path("tft_bot/config.yaml")
+    config_path = f"{storage_path}\\config.yaml"
+
+    if not os.path.isfile(config_path):
+        shutil.copyfile(config_resource_path, config_path)
+
+    with open(config_resource_path, mode="r", encoding="UTF-8") as config_resource:
+        _config_resource: dict[str, Any] = yaml.load(config_resource)
+
+    with open(config_path, mode="r", encoding="UTF-8") as config_file:
+        global _SELF
+        _SELF = yaml.load(config_file)
+
+    if _config_resource.get("version") > _SELF.get("version", 0):
+        logger.warning("Config is outdated, creating a back-up and updating it")
+
+        shutil.copyfile(config_path, f"{config_path}.bak")
+
+        _config_resource.update((key, _SELF[key]) for key in _SELF.keys() & _config_resource.keys() if key != "version")
+        with open(config_path, mode="w", encoding="UTF-8") as config_file:
+            yaml.dump(_config_resource, config_file)
+
+        _SELF = _config_resource
+
+    parse_cli_flags()
+
+
+def parse_cli_flags() -> None:
+    """
+    Override settings for the bot with CLI flags, since they take higher precedence.
+    """
+    arg_parser = argparse.ArgumentParser(prog="TFT Bot")
+    arg_parser.add_argument(
+        "-f",
+        "--ffearly",
+        action="store_true",
+        help="If the game should be surrendered at first available time.",
+    )
+    arg_parser.add_argument(
+        "-v",
+        "--verbose",
+        action="store_true",
+        help="Increase output verbosity, mostly useful for debugging",
+    )
+    parsed_args = arg_parser.parse_args()
+
+    if parsed_args.ffearly:
+        _CLI_FLAGS["FORFEIT_EARLY"] = True
+
+    if parsed_args.verbose:
+        _CLI_FLAGS["VERBOSE"] = True
```

Comment:
I like the idea that we're not altering the bot's copy of the parsed YAML config, _however_ is there any reason to actually keep the `_CLI_FLAGS` separate from the `_SELF`? The only scenario I can think of is when the bot would try to re-write the file with the changed configs, but is there a reason that would happen?

My main question/concern is just for the readability below.

I don't feel super strongly about this, it's really just a question I have in my head that I haven't decided yet and wanted to get your opinion on since you went this route!

Reply:
Hmm, there isn't a big reason to separate them. I've done it for my mind to actively think about implementing it, but since we've documented them as genuinely overriding it, it'd be fine to override the values in _SELF after they are loaded. Even if the config were to be re-loaded for whichever reason, it'd load it from the file first before overriding it again, so there'd be no chance of it being written to the file. I'll change it!